src/lib/poptrie_lookup.dasl

```diff
@@ -0,0 +1,158 @@
+-- Use of this source code is governed by the Apache 2.0 license; see COPYING.
+
+module(..., package.seeall)
+
+local debug = false
+
+local ffi = require("ffi")
+local dasm = require("dasm")
+
+|.arch x64
+|.actionlist actions
+|.globalnames globalnames
+
+-- Table keeping machine code alive to the GC.
+local anchor = {}
+
+-- Assemble a lookup routine
+function generate (Poptrie, keysize)
+   -- Assert assumptions about lib.poptrie
+   assert(Poptrie.k == 6)
+   if Poptrie.direct_pointing then
+      assert(Poptrie.leaf_tag == bit.lshift(1, 31))
+   end
+   assert(ffi.sizeof(Poptrie.leaf_t) == 2)
+   assert(ffi.sizeof(Poptrie.vector_t) == 8)
+   assert(ffi.sizeof(Poptrie.base_t) == 4)
+   assert(ffi.offsetof(Poptrie.node_t, 'leafvec') == 0)
+   assert(ffi.offsetof(Poptrie.node_t, 'vector') == 8)
+   assert(ffi.offsetof(Poptrie.node_t, 'base0') == 16)
+   assert(ffi.offsetof(Poptrie.node_t, 'base1') == 20)
+
+   local name = "poptrie_lookup(k="..Poptrie.k..", keysize="..keysize..")"
+
+   local Dst = dasm.new(actions)
+   lookup(Dst, Poptrie, keysize)
+   local mcode, size = Dst:build()
+   table.insert(anchor, mcode)
+
+   if debug then
+      print("mcode dump: "..name)
+      dasm.dump(mcode, size)
+   end
+
+   local prototype
+   if keysize <= 64 then
+      prototype = ffi.typeof(
+         "$ (*) ($ *, $ *, uint64_t, $ *)",
+         Poptrie.leaf_t, Poptrie.leaf_t, Poptrie.node_t, Poptrie.base_t
+      )
+   else error("NYI") end
+
+   return ffi.cast(prototype, mcode)
+end
+
+|.define leaves,  rdi -- pointer to leaves array
+|.define nodes,   rsi -- pointer to nodes array
+|.define key,     rdx -- key to look up
+|.define dmap,    rcx -- pointer to directmap
+|.define index,   r8d -- index into node array
+|.define node,    r8  -- pointer into node array 
+|.define offset,  r9  -- offset into key
+|.define v,       r10 -- k or s bits extracted from key
+|.define vec,     r11 -- 64-bit vector or leafvec
+
+-- lookup(leaf_t *leaves, node_t *nodes, key) -> leaf_t
+function lookup (Dst, Poptrie, keysize)
+   if Poptrie.direct_pointing then
+      -- v = extract(key, 0, Poptrie.s)
+      local direct_mask = bit.lshift(1ULL, Poptrie.s) - 1
+      -- v = band(key, direct_mask)
+      | mov v, direct_mask
+      | and v, key
+      -- index = dmap[v]
+      | mov index, dword [dmap+v*4]
+      -- eax = band(index, leaf_tag)
+      | mov eax, 0x80000000
+      | and eax, index
+      -- is leaf_tag set?
+      | cmp eax, 0
+      | je >1 -- leaf_tag not set, index is a node
+      -- eax = leaf_tag - 1 (tag inverted)
+      | mov eax, 0x7FFFFFFF
+      | and eax, index
+      | ret
+      -- node, offset = nodes[index], s
+      |1:
+      | imul index, 24 -- multiply by node size
+      | lea node, [nodes+index]
+      -- offset = s
+      | mov offset, 18
+   else
+      -- index, node, offset = 0, nodes[index], 0
+      | mov index, 0
```

Comment:
Unless you're explicitly preserving flags (which I assume you're not), `xor R, R` is typically better than `mov R, 0` (e.g. because it is a byte or two shorter to encode).